### PR TITLE
[FSDP] Fix pre-hook check when module is only used in forward

### PR DESCRIPTION
**Pull request by zachzzc**
In a corner case, a sub-module is only used in forward but not in backward. The assert_state will fail because the module is still in BACKWARD_IDLE state. Add condition to make sure the module has run in backward.
@ronghanghu @JackCaoG

### CODE REVIEW

**Comment by ronghanghu** [comment]
Thanks @zachzzc for the PR! Hope to consult you on the further details:

> The assert_state will fail because the module is still in BACKWARD_IDLE state. Add condition to make sure the module has run in backward.

Do you mean `BACKWARD_PRE` state or `IDLE` state? (There isn't a `BACKWARD_IDLE` state. I guess you meant `BACKWARD_PRE` state?).

In addition
- I'm thinking we still want to call `_finalize_parameters` on such modules. Maybe we could move the check on `m._pre_backward_hook_has_run` to one of the branches in https://github.com/pytorch/xla/blob/1d313bb08ade4240ffd5ac8d6775264cfddacfa7/torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py#L1271-L1274
- I'm thinking in this case one should also call `_free_full_params` on those parameters.

(Ideally in this case it would be better if we can prevent `_pre_backward_hook` on such modules from firing, so that its full parameters are never built during the backward pass. We probably had a similar issue in https://github.com/pytorch/xla/pull/4485#discussion_r1110304862 during FSDP evaluation, where somehow the hook probably messed up during model evaluation without a backward pass.)

**Comment by alanwaketan** [comment]
Yea, I guess it's better to investigate what Ronghang suggests instead of relaxing the check.

**Comment by zachzzc** [comment]
@alanwaketan @ronghanghu  Ready for review

**Review by alanwaketan — COMMENTED** [review]
Also, can you verify that this is can remove the `torch.no_grad` context in the following test?
https://github.com/pytorch/pytorch/blob/0466784dcf8addb8bb84d253ca23e695f22c27b5/test/test_train_mp_imagenet_fsdp.py#L358

File: test/test_fsdp_auto_wrap.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+  def test(self):
+    dev = xm.xla_device()
+    input = torch.zeros([16, 16], device=dev)
+    model = self.MyModel(input_size=16, hidden_size=4)
+    model = XlaFullyShardedDataParallel(
+        model, auto_wrap_policy=always_wrap_policy)
+    model.to(dev)
+    hid1, hid2 = model(input)
+    loss = hid1.sum()
```
[the remark this answers is not in the document]
Comment (zachzzc) [reply test/test_fsdp_auto_wrap.py:39]: One linear in the module is used only in the forward but not in the backward path. We would see the assertion error without this PR.

[the remark this answers is not in the document]
Comment (alanwaketan) [reply test/test_fsdp_auto_wrap.py:39]: Maybe a comment in the test would be great.

File: torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -891,7 +891,7 @@ def forward(self, *args: Any, **kwargs: Any) -> torch.Tensor:
       # This can be used to debug FSDP parameter memory consumption.
       outputs = self._dummy_forward(*args, **kwargs)
 
-    if self.reshard_after_forward:
+    if self.reshard_after_forward or not self.training:
```
Comment (ronghanghu) [inline torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py:894]: A concern regarding `self.training` change here and below (and the change in test/test_train_mp_imagenet_fsdp.py) -- I feel we should still allow users to call backward on a model in eval mode, but with the change here, such cases would crash.

I think there could be legitimate use cases where the user want to do back-propagation in such cases (e.g. one of them is [domain adaptation](https://stackoverflow.com/questions/57323023/pytorch-loss-backward-and-optimizer-step-in-eval-mode-with-batch-norm-laye) and there are also cases mentioned on PyTorch [forum](https://discuss.pytorch.org/t/model-eval-vs-with-torch-no-grad/19615)).

Comment (zachzzc) [reply torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py:894]: This is an interesting use cases that I didn't think of. If that's the case, I didn't have any better solution to address @alanwaketan 's TODO, will revert it back.

**Review by alanwaketan — APPROVED** [review]
LGTM. Thanks, Zack.

**Comment by zachzzc** [comment]
@ronghanghu @alanwaketan could you help merge the PR it everything looks good to you?

**Review by ronghanghu — APPROVED** [review]
